File: 417-pacific-atlantic-water-flow/pacific-atlantic-water-flow.cpp
class Solution {
public:
    int n, m;
    vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    void bfs(queue<pair<int, int>>& q, vector<vector<bool>>& vis, vector<vector<int>>& grid) 
    {
        while (!q.empty()) 
        {
            auto [row, col] = q.front();
            q.pop();
        
            for (auto dir : directions) 
            {
                int newRow = row + dir.first;
                int newCol = col + dir.second;

                // Check boundaries and if already visited
                if (newRow >= 0 && newRow < n && newCol >= 0 && newCol < m && 
                !vis[newRow][newCol] && grid[row][col] <= grid[newRow][newCol]) 
                {
                    vis[newRow][newCol] = true;
                    q.push({newRow, newCol});
                }
            }
        }
    }

    vector<vector<int>> pacificAtlantic(vector<vector<int>>& grid) 
    {
        n = grid.size();
        m = grid[0].size();

        queue<pair<int, int>> pacific, atlantic;
        vector<vector<bool>> pacificVis, atlanticVis;
        atlanticVis = pacificVis = vector<vector<bool>>(n, vector<bool>(m, false));
        // start bfs from the shore of pacific and atlantic
        for(int i = 0; i < m; i++)
        {
            pacific.push({0, i});
            atlantic.push({n - 1, i});
            pacificVis[0][i] = true;
            atlanticVis[n - 1][i] = true;
        }

        for(int i = 0; i < n; i++)
        {
            pacific.push({i, 0});
            atlantic.push({i, m - 1});
            pacificVis[i][0] = true;
            atlanticVis[i][m - 1] = true;
        }

        // start bfs from the shore of atlantic and pacific
        bfs(pacific, pacificVis, grid);
        bfs(atlantic, atlanticVis, grid);

        vector<vector<int>> res;
        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < m; j++)
            {
                if(pacificVis[i][j] && atlanticVis[i][j])
                res.push_back({i, j});
            }
        }
        return res;
    }
};

File: 417-pacific-atlantic-water-flow/pacific-atlantic-water-flow.java
class Solution {
    private int n, m;
    private final int[][] directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    private boolean isValidCell(int row, int col){
        return row >= 0 && row < n && col >= 0 && col < m;
    }
    
    private void BFS(int[][] heights, boolean[][] visited, Queue<int[]> queue){
        while(!queue.isEmpty()){
            int[] curr = queue.poll();
            int i = curr[0];
            int j = curr[1];
            for(int[] d : directions){
                int row = d[0] + i;
                int col = d[1] + j;
                if(isValidCell(row, col) && !visited[row][col] && heights[i][j] <= heights[row][col]){
                    visited[row][col] = true;
                    queue.offer(new int[]{row, col});
                }
            }
        }
    }
    
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        n = heights.length;
        m = heights[0].length;
        boolean[][] pacific = new boolean[n][m];
        boolean[][] atlantic = new boolean[n][m];

        Queue<int[]> queue = new LinkedList<>();
        for(int i = 0; i < n; i++){
            queue.offer(new int[]{i, 0});
            pacific[i][0] = true;
        }

        for(int j = 0; j < m; j++){
            queue.offer(new int[]{0, j});
            pacific[0][j] = true;
        }

        BFS(heights, pacific, queue);
        queue.clear();

        for(int i = 0; i < n; i++){
            queue.offer(new int[]{i, m - 1});
            atlantic[i][m - 1] = true;
        }

        for(int j = 0; j < m; j++){
            queue.offer(new int[]{n - 1, j});
            atlantic[n - 1][j] = true;
        }

        BFS(heights, atlantic, queue);
        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++){
                if(pacific[i][j] && atlantic[i][j]){
                    res.add(List.of(i, j));
                }
            }
        }
        return res;
    }
}